lib/SIL/InstructionUtils.cpp

```diff
@@ -530,3 +532,137 @@ void swift::findClosuresForFunctionValue(
+    oldOperandTypes = oldOperandTypes.drop_front();
+  } else {
+    resultType = oldResultType;
+  }
+
+  // Then go through all of our values and bail if any after substitution are
+  // not concrete builtin types. Otherwise, stash each of them in the argTypes
+  // array as objects. We will convert them as appropriate.
+  for (SILType ty : oldOperandTypes) {
+    // If after specialization, we do not have a builtin type, bail.
+    if (!ty.is<BuiltinType>())
+      return false;
+
+    // Otherwise, we have an object builtin type ready to go.
+    argTypes.push_back(ty.getObjectType());
+  }
+
+  // Ok, we have all builtin types. Infer the underlying polymorphic builtin
+  // name form our first argument.
+  CanBuiltinType builtinType = argTypes.front().getAs<BuiltinType>();
+  SmallString<32> builtinTypeNameStorage;
+  StringRef typeName = builtinType->getTypeName(builtinTypeNameStorage, false);
+  staticOverloadName.append("_");
+  staticOverloadName.append(typeName);
+
+  auto &ctx = mod.getASTContext();
+  staticOverloadIdentifier = ctx.getIdentifier(staticOverloadName);
+  return true;
+}
+
+bool PolymorphicBuiltinSpecializedOverloadInfo::init(BuiltinInst *bi) {
+#ifndef NDEBUG
+  assert(!isInitialized && "Can not init twice?!");
+  SWIFT_DEFER { isInitialized = true; };
+#endif
+
+  // First quickly make sure we have a /real/ BuiltinValueKind, not an intrinsic
+  // or None.
+  auto kind = bi->getBuiltinKind();
+  if (!kind)
+    return false;
+
+  SmallVector<SILType, 8> oldOperandTypes;
+  copy(bi->getOperandTypes(), std::back_inserter(oldOperandTypes));
+  assert(bi->getNumResults() == 1 &&
+         "We expect a tuple here instead of real args");
+  SILType oldResultType = bi->getResult(0)->getType();
+  return init(bi->getModule(), *kind, oldOperandTypes, oldResultType);
+}
+
+SILValue
+swift::getStaticOverloadForSpecializedPolymorphicBuiltin(BuiltinInst *bi) {
+
+  PolymorphicBuiltinSpecializedOverloadInfo info;
+  if (!info.init(bi))
+    return SILValue();
+
+  SmallVector<SILValue, 8> rawArgsData;
+  copy(bi->getOperandValues(), std::back_inserter(rawArgsData));
+
+  SILValue result = bi->getResult(0);
+  MutableArrayRef<SILValue> rawArgs = rawArgsData;
+
+  if (info.hasOutParam) {
+    result = rawArgs.front();
+    rawArgs = rawArgs.drop_front();
+  }
+
+  assert(bi->getNumResults() == 1 &&
+         "We assume that builtins have a single result today. If/when this "
+         "changes, this code needs to be updated");
+
+  SILBuilderWithScope builder(bi);
+  // Ok, now we know that we can convert this to our specialized
+  // builtin. Prepare the arguments for the specialized value, loading the
+  // values if needed at +1 and storing the result into an out parameter if
+  // needed.
+  SmallVector<SILValue, 8> newArgs;
+  for (auto arg : rawArgs) {
+    if (arg->getType().isObject()) {
+      newArgs.push_back(arg);
+      continue;
+    }
+
+    SILValue load = builder.emitLoadValueOperation(
```

Comment:
I am going to look into tightening this up a little bit in a forthcoming commit. This should really be loading a trivial value to make it clear that the +1 doesnt matter.

Reply:
Actually found a reason to stop the testing. I am going to do it in this PR.